File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.py
import sys
from collections import deque
input=sys.stdin.readline

def find(n,nx,ny,mx,my):
  if nx==mx and ny==my:
    return 0

  visited=[[False for i in range(n)]for j in range(n)]
  q=deque([[nx,ny,0]])
  while q:
    [x,y,now]=q.popleft()
    if x<0 or y<0 or x>=n or y>=n or visited[x][y]:
      continue
    if x==mx and y==my:
      return now
    visited[x][y]=True
    q.append([x-2,y+1,now+1])
    q.append([x-2,y-1,now+1])
    q.append([x-1,y+2,now+1])
    q.append([x-1,y-2,now+1])
    q.append([x+2,y+1,now+1])
    q.append([x+2,y-1,now+1])
    q.append([x+1,y+2,now+1])
    q.append([x+1,y-2,now+1])


T=int(input())
for i in range(T):
  n=int(input())
  nowx,nowy=map(int,input().split())
  movex,movey=map(int,input().split())
  print(find(n,nowx,nowy,movex,movey))

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.js
const readline=require('readline')
const rl=readline.createInterface({
  input:process.stdin,
  output:process.stdout
})

const input=[]
let iter=0
rl.on('line',(line)=>{
  input.push(line)
}).on('close',()=>{
  const T=Number(input[iter++])
  const answer=[]
  for(let _=0;_<T;_++){
    const N=Number(input[iter++])
    const [nx,ny]=input[iter++].split(' ').map(Number)
    const [mx,my]=input[iter++].split(' ').map(Number)

    if (nx===mx && ny===my) {
      answer.push(0)
      continue
    }

    const visited=[...new Array(N)].map(item=>[...new Array(N)].map(item=>false))
    visited[nx][ny]=true

    const xbox=[1,2,2,1,-1,-2,-2,-1]
    const ybox=[-2,-1,1,2,-2,-1,1,2]

    const q=[[nx,ny,0]]
    let finish=false
    while (q.length>0){
      const [nx,ny,cnt]=q.shift()
      for(let i=0;i<8;i++){
        const newx=nx+xbox[i]
        const newy=ny+ybox[i]
        if (0<=newx && newx<N && 0<=newy && newy<N){
          if (visited[newx][newy]) continue
          visited[newx][newy]=true
          if (newx===mx && newy===my){
            answer.push(cnt+1)
            finish=true
            break
          }
          q.push([newx,newy,cnt+1])
        }
      }
      if (finish) break
    }
  }

  console.log(answer.join('\n'))
  process.exit()
})
